pnpm/test/monorepo/index.ts

```diff
@@ -1855,3 +1855,126 @@ shared-workspace-lockfile=false
+  const preparedManifests = {
+    root: {
+      name: 'root',
+      version: '0.0.0',
+      private: true,
+      pnpm: {
+        neverBuiltDependencies: [],
+      },
+    },
+    'project-0': {
+      name: 'project-0',
+      version: '0.0.0',
+      dependencies: {
+        '@pnpm.e2e/install-script-example': '1.0.0',
+      },
+    },
+  } satisfies Record<string, ProjectManifest>
+
+  preparePackages([
+    {
+      location: '.',
+      package: preparedManifests.root,
+    },
+    preparedManifests['project-0'],
+  ])
+
+  fs.writeFileSync('.npmrc', `
+inject-workspace-packages=true
+`, 'utf8')
+  writeYamlFile('pnpm-workspace.yaml', { packages: ['**', '!store/**'] })
+
+  const monorepoRoot = process.cwd()
+  const deployOutputProjectDir = path.join(makeTempDir(false), './project-0-deployed')
+
+  {
+    const { status } = execPnpmSync(['install'])
+    expect(status).toBe(0)
+  }
+  {
+    process.chdir('project-0')
+    expect(fs.existsSync('node_modules/@pnpm.e2e/install-script-example/generated-by-install.js')).toBeTruthy()
+  }
+  {
+    process.chdir(monorepoRoot)
+    const { status } = execPnpmSync(['deploy', '--filter', 'project-0', deployOutputProjectDir])
+    expect(status).toBe(0)
+  }
+  {
+    process.chdir(deployOutputProjectDir)
+    expect(fs.existsSync('node_modules/@pnpm.e2e/install-script-example/generated-by-install.js')).toBeTruthy()
+  }
+})
+
+test('rebuild in a directory created with "pnpm deploy" and with "pnpm.neverBuiltDependencies" configured should run lifecycle scripts', async () => {
+  const preparedManifests = {
+    root: {
+      name: 'root',
+      version: '0.0.0',
+      private: true,
+      pnpm: {
+        neverBuiltDependencies: [],
+      },
+    },
+    'project-0': {
+      name: 'project-0',
+      version: '0.0.0',
+      dependencies: {
+        '@pnpm.e2e/install-script-example': '1.0.0',
+      },
+    },
+  } satisfies Record<string, ProjectManifest>
+
+  preparePackages([
+    {
+      location: '.',
+      package: preparedManifests.root,
+    },
+    preparedManifests['project-0'],
+  ])
+
+  fs.writeFileSync('.npmrc', `
+inject-workspace-packages=true
+`, 'utf8')
+  writeYamlFile('pnpm-workspace.yaml', { packages: ['**', '!store/**'] })
+
+  const monorepoRoot = process.cwd()
+  const deployOutputProjectDir = path.join(makeTempDir(false), './project-0-deployed')
+
+  {
+    const { status } = execPnpmSync(['install'])
+    expect(status).toBe(0)
+  }
+  {
+    process.chdir('project-0')
+    expect(fs.existsSync('node_modules/@pnpm.e2e/install-script-example/generated-by-install.js')).toBeTruthy()
+  }
+  {
+    process.chdir(monorepoRoot)
+    const { status } = execPnpmSync(['deploy', '--filter', 'project-0', deployOutputProjectDir])
+    expect(status).toBe(0)
+  }
+  {
+    process.chdir(deployOutputProjectDir)
+    fs.rmSync('node_modules/@pnpm.e2e/install-script-example/generated-by-install.js')
```

Comment:
@KSXGitHub could you help me with this test case?

This `rmSync` currently throws because the file does not exist.  
But I would expect that here - after `pnpm deploy` - the file `node_modules/@pnpm.e2e/install-script-example/generated-by-install.js` exists.  
It seems like `pnpm deploy` did not correctly copy-over that file created by a lifecycle script!

Manually, I cannot reproduce this issue.

Reply:
https://github.com/pnpm/pnpm/pull/8990#issuecomment-2614513135